pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +214,195 @@ pub mod pallet {
+				for (currency_id, _) in total_issuances {
+					T::Fungibles::start_destroy(currency_id, None)?;
+				}
+				Self::deposit_event(Event::RefundCompleted(pool_id.clone()));
+			} else {
+				// deactivate all currencies whose issuance dropped to 0 in this step
+				for currency_id in dead_currencies {
+					T::Fungibles::start_destroy(currency_id.clone(), None)?;
+				}
+			}
+
+			Ok(())
 		}
 
 		#[pallet::call_index(9)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn finish_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn finish_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::InUse);
+
+			for currency_id in pool_details.bonded_currencies {
+				if T::Fungibles::asset_exists(currency_id.clone()) {
+					// This would fail with an InUse error if there are any accounts left on any currency
+					T::Fungibles::finish_destroy(currency_id.clone())?;
+				}
+			}
+
+			let pool_account = pool_id.clone().into();
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			if !total_collateral_issuance.is_zero() {
+				T::CollateralCurrency::transfer(
+					T::CollateralAssetId::get(),
+					&pool_account,
+					&pool_details.manager,
+					total_collateral_issuance,
+					Preservation::Expendable,
+				)?;
+			}
+
+			Pools::<T>::remove(&pool_id);
+
+			// TODO: refund deposit
+
+			Self::deposit_event(Event::Destroyed(pool_id));
+
+			Ok(())
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		fn do_start_destroy_pool(pool_id: T::PoolId, maybe_check_owner: Option<AccountIdOf<T>>) -> DispatchResult {
+			let mut pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(!pool_details.state.is_destroying(), Error::<T>::Destroying);
+
+			if let Some(caller) = maybe_check_owner {
+				ensure!(pool_details.is_manager(&caller), Error::<T>::Unauthorized);
+			}
+
+			pool_details.state.destroy();
+
+			Self::deposit_event(Event::DestructionStarted(pool_id.clone()));
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_id.clone().into());
+
+			if total_collateral_issuance.is_zero() {
+				// deactivate all currencies and emit RefundCompleted
+				for currency_id in pool_details.bonded_currencies.iter() {
+					T::Fungibles::start_destroy(currency_id.clone(), None)?;
+				}
+
+				Self::deposit_event(Event::RefundCompleted(pool_id.clone()));
+
+				return Ok(());
+			}
+
+			let mut destroyed: usize = 0; 
+			for currency_id in pool_details.bonded_currencies.iter() {
+				if T::Fungibles::total_issuance(currency_id.clone()).is_zero() {
+					T::Fungibles::start_destroy(currency_id.clone(), None)?;
```

Comment:
This function should never fail if you know the currency ID exists. I would suggest to add an `Internal` error and throw that here, perhaps with a `log::error` that logs the inner error.

Reply:
What's the reason not to just propagate the dispatch error produced by the underlying implementation, if there is any? Why wrap it with an opaque `InternalError` that's going to be useless for anyone trying to figure out what went wrong?

Reply:
Because this is never supposed to be thrown in the first place. If it does, there's something worse going on, and we should treat it with urgency. Every time we see an `Internal` error being generated we know it's bad. If we bubble up the inner error, there are definitely cases in which such error could be expected, hence we need to always check if it needs investigation. I guess at the very least you should `log::error` it so that we have it in the logs. But the logs can only go so far if we don't know we need to check them. So `Internal` is a big red flag to indicate something potentially wrong, like some broken invariants, is going on with the system.

Reply:
But the 'broken invariant' assumption is based on the current trait implementation in the assets pallet, which indeed would never fail, even if the asset is already in destroying state. In my mind it goes against the whole idea of loose coupling to base your assumptions on the implementation of a single pallet where the trait itself makes no guarantees that this works (maybe some preconditions for currency destruction are not met?). So bubbling up potential errors of that underlying implementation makes a lot more sense to me than masking it with some 'InternalError' which then means you'd have to look at the logs to figure out what's wrong, which most people obviously don't have access to.

Reply:
There could be other stuff going on in the assets pallet that are outside the control of this pallet, for instance operating on the currency itself via that pallet without going through this pallet. It could be a governance action that we did not foresee could have implications on a different, loosely connected pallet such as this one. 
In general, logs are generated from the runtime, so replaying the failing block with chopsticks is enough to access them and debug the issue. In general, assuming that stuff should not fail and throw an `Internal` error if they do, is the equivalent, in my opinon, to calling `.uwnrap` or `.expect`. This said, you're free to leave the error as it is, perhaps after @Ad96el also provides his opinion on this. For me it's more a matter of making broken assumption more clearly visible than otherwise they would be, whatever the reason for them to break is.

Reply:
I have to agree, I missed that the operation should never fail. As Antonio mentioned, when an operation should never fail, it's a broken invariant. The internal error isn’t meant for the user but for us to understand what went wrong under the hood, using logs.

Since we agreed to introduce logs later, using an internal error seems like the right approach for now.

Reply:
I have to say, the more I think about it, the less I like the idea of just masking any errors from the underlying implementation. InternalErrors are fine when it concerns internal logic that seems flawed, but if it's a external code that simply does not behave like the assets pallet would, I don't think that's a good idea

Reply:
In web development if the database drops, the website would probably return a `500`, and as a developer as soon as you see `500` you know that something is wrong and needs further investigation. Now assume it's not a DB but it's an external service that returns `404`, when it really should not, you most likely would not return `404` back to the user, but rather a `5xx` error code, right? Since it's not your service that broke, but a dependency. I see the `Internal` error working in the same way. Anyway, if you decide to go ahead this way, just make sure you and @Ad96el find an agreement.